pkg/networkservice/common/heal/client.go

```diff
@@ -88,25 +89,72 @@ func (u *healClient) Request(ctx context.Context, request *networkservice.Networ
 
 	connInfo, loaded := u.conns.LoadOrStore(conn.GetId(), &connectionInfo{
 		conn: conn.Clone(),
-		mut:  sync.Mutex{},
+		cond: sync.NewCond(new(sync.Mutex)),
 	})
 	u.replaceConnectionPath(conn, connInfo)
 
-	conn, err := next.Client(ctx).Request(ctx, request, opts...)
-	if err != nil {
-		if !loaded {
+	defer func() {
+		if !loaded && err != nil {
 			u.conns.Delete(request.GetConnection().GetId())
 		}
+	}()
+
+	conn, err = next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
 		return nil, err
 	}
 
-	connInfo.mut.Lock()
-	defer connInfo.mut.Unlock()
+	condCh, cancel := u.condCh(connInfo)
+	defer cancel()
+
+	// Problem:
+	//   gRPC requests and streaming requests have different requirements for the underlying gRPC connection. There can
+	//   be a case on the closing gRPC connection, when it is still sending requests but fails to keep or start
+	//   a streaming request. In such case we can get a situation when we have a Connection with no monitor stream, so
+	//   it would never be healed.
+	// Solution:
+	//   We should check that monitor stream is ready for the Connection, before returning it back. This check should
+	//   wait for some time after the next.Request(), because for the [NSMgr -> Endpoint] case response may come faster
+	//   than monitor stream update.
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-condCh:
+	case <-time.After(100 * time.Millisecond):
```

Comment:
we should probably use clocktime.After here

Reply:
Yes, sure, thank you for the notice :)